### DISCUSSION TITLE
Fargate - Trouble getting files bundled while using dynamic payload paths based on environment flag passed from CLI

### DISCUSSION THREAD

**Discussion by DanSmithSC** [artilleryio/artillery#2281] [Artillery]
While trying to run a Fargate test from a config which takes a dynamic -environment flag in as a parameter, I'm having difficulty getting the required files to be bundled resulting in some `ENOENT: no such file or directory, open` errors.

While working with multiple environments I'm passing my environment in as flag like such from the root of the project directory:
`$ artillery run-fargate --region us-east-1 -e qa ./artillery/<app-to-test>.yml`

The App to Test is a specific config file relevant to the application with routes to be tested and scenarios/flows.

The environment flag in this case `qa` is being read in dynamically appropriately. It is then used in the config as a variable referenced with `{{ environment }}`. However, paths that rely on that dynamic environment flag are not being bundled appropriately and are not found when using the Fargate implementation.

From the Project file hierarchy is structured as such with the required CSVs to be read in under their respective environment folder that is in the `/config` directory but also CSVs that are relevant for all environments.
```
 /Root
 - /artillery
    --app-1.yml
    --app-2.yml
    --app-3.yml
 
 - /config
   --/dev
      - file1.csv
      - file2.csv
      - <...>
   --/qa
      - file1.csv
      - file2.csv
      - <...>
   --/stage
      - file1.csv
      - file2.csv
      - <...>
  -data.csv    

<...>      
```
When attempting to access these paths in the config in the config.payload.paths dynamically, the test will fail as the files are not bundled during the provisioning and setup of the containers:

`app-to-test.yml`
```
config:
 <...>
  includeFiles: 
    - "../config/{{ $environment }}/file1.csv"
    - "../config/{{ $environment }}/file2.csv"
    - "../config/{{ $environment }}/file3.csv"
    - "../config/{{ $environment }}/file4.csv"
    - "../config/data.csv"    
  payload:
    -
      path: "../config/{{ $environment }}/file1.csv"
      delimiter: "\n"
     <...>
    -
      path: "../config/{{ $environment }}/file2.csv"
      delimiter: "\n"
      <...>
    -
      path: "../config/{{ $environment }}/file3.csv"
      delimiter: "\n"
      <...>
    -
      path: "../config/{{ $environment }}/file4.csv"
      delimiter: "\n"
      <...>
    -
      path: "../config/data.csv"
      delimiter: "\n"
      <...>
```

When Fargate does its bundling of files it appears to not appropriately bundle the nested environment's folder that contains the CSVs relevant to that environment, but it WILL bundle the `data.csv` file that is located directly under the `/config` folder.

I can see in my task logs the following has been logged as pre-packaged/included:
```
.
| -- artillery
| `-- app-1.yml`
| -- config
| `-- data.csv`
| --metadata.json
| --package-lock.json
| --package.json
2 directories, 5 files
```
I would hope to see something like the following should all the files be read in as expected:
```
.
| -- artillery
| `-- app-1.yml`
| -- config
| `--qa
|  ``-- file1.csv
|  ``-- file2.csv
|  ``-- file3.csv
|  ``-- file4.csv
| `-- data.csv`
| --metadata.json
| --package-lock.json
| --package.json
3 directories, 9 files
```

After printing the CLI args it attempts to run the node process but Errors with the following:
```
Error: ENOENT: no such file or directory, open
'/artillery-pro/test_data/config/qa/file1.csv'
```

So i'm wondering if there's an issue with how Artillery is bundling up and determining the dependencies and files required based on the config structure when it's being read in dynamically from the passed in `-e` flag arg.

I've tried searching for topics or issues around how it determines the files to bundle, but couldn't quite find anything that helped. I attempted to use the `includeFiles` flag at the top of the config, but that didn't seem to make any difference. Thanks in advance!

**Comment by bernardobridge** [comment]
Hi @DanSmithSC 👋 ,

Currently, the `includeFiles` option (part of Fargate) cannot receive dynamic values, as it is not templated. That will change in the near future with some work we have planned to improve the Fargate setup, but it's a known limitation right now.

I believe if you specify each file individually in your `includeFiles` with static values, Fargate should correctly bundle them all. This is why you see `config/data.csv`, but not the others. Then, in each worker, Artillery should be able to find the relevant file based on the dynamically set `$environment` in the payload path.

Hope that helps!

**Reply by DanSmithSC** [reply]
Thanks for the response! - I ended up trying a hard-coded static `includeFiles` reference for the environment and it did run the test and find the files and bundle them. I'll keep an eye out for the future enhancements work to come!
